### ISSUE TITLE
Uri: add support for parsing URLs (suffix references)

### ISSUE THREAD

**Issue by har79** [dart-lang/sdk#33817]
Dart SDK 2.0.0-dev.63.0 (DartPad)

`Uri.parse` treats [suffix references](https://tools.ietf.org/html/rfc3986#section-4.5), e.g., "www.google.com", as the path component which can result in the output being incorrect.

```dart
final uri = Uri.parse('www.dartlang.org/install');

print('authority: ${uri.authority}'); // authority: 
print('path: ${uri.path}'); // path: www.dartlang.org/install
print(uri.replace(scheme: 'http')); // http:www.dartlang.org/install
```

Although suffix references don't conform to correct URI syntax, so there's nothing technically incorrect about this behaviour, it would be very convenient to have some support given how common they are. This could either be through heuristics or a second parse method, e.g., a `parseUrl` that treats the initial component of a schema-less string as the authority:

```dart
final uri = Uri.parseUrl('www.dartlang.org/install');

print('authority: ${uri.authority}'); // authority: www.dartlang.org
print('path: ${uri.path}'); // path: /install
print(uri.replace(scheme: 'http')); // http://www.dartlang.org/install
```

**Comment by lrhn** [comment]
It will definitely have to be a separate parse function.
It's fairly easy to do, actually:
```dart
Uri parseSuffixReference(String input) => Uri.parse("//$input");
```
If you want to parse heuristically, then maybe:
```dart
Uri parsePotentialSuffixReference(String input) {
  if (input.startsWith(_probablyHostNameRE)) input = "//$input";
  return Uri.parse(input);
}
final _probablyHostNameRE = RegExp(r"\w+(\.\w+)+/");
```

I'm not sure I think this use-case is general enough to warrant a function on the `Uri` class.

**Comment by har79** [comment]
Yeah, I was looking for something like that but that also handles if the scheme is specified, i.e., would cope with both "http://google.com" and "google.com". Similar to:

```
Uri parseUrl(String input) {
  final uri = Uri.parse(input);
  return uri.hasAuthority ? uri : Uri.parse('http://$input');
}
```

but without the inefficiency of parsing twice.

**Comment by lrhn** [comment]
I believe the `parsePotentialSuffixeReference` would do that since a scheme must end with `:`, so it doesn't match the regexp. It would parse `http://x/y/z` normally and `x.y/z` as `//x.y/z`.

You can argue, though, that the inefficiency of parsing twice is still there since it does a RegExp match on the string before parsing.
